src/os_common/os_fhandle.c

```diff
@@ -355,6 +355,101 @@ __wt_close(WT_SESSION_IMPL *session, WT_FH **fhp)
 	return (__handle_close(session, fh, true));
 }
 
+/*
+ * __wt_fsync_all_background --
+ *	POSIX background fsync for all dirty file handles.
+ */
+int
+__wt_fsync_all_background(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_FH *fh, *fhtmp;
+	WT_FILE_HANDLE *handle;
+	uint64_t now;
+	const char *fp, *ftp;
+
+	conn = S2C(session);
+	WT_ASSERT(session, !F_ISSET(conn, WT_CONN_READONLY));
+	WT_STAT_CONN_INCR(session, fsync_all);
+	__wt_spin_lock(session, &conn->fh_lock);
+	TAILQ_FOREACH_SAFE(fh, &conn->fhqh, q, fhtmp) {
+		handle = fh->handle;
+		WT_STAT_CONN_INCR(session, fsync_all_fh_total);
+		if (!F_ISSET(fh, WT_FH_DIRTY) ||
+		    handle->fh_sync_nowait == NULL ||
+		    fh->written < WT_CAPACITY_FILE_THRESHOLD) {
+			WT_STAT_CONN_INCR(session, fsync_too_small);
+			continue;
+		}
+		/* Skip over WiredTiger owned files. */
+		fp = strrchr(fh->name, '/');
+		if (fp == NULL)
+			fp = fh->name;
+		else
+			++fp;
+		if (fp[0] == 'W' &&
+		    WT_PREFIX_MATCH(fp, "WiredTiger"))
+			continue;
+		now = __wt_clock(session);
+		if (fh->last_sync == 0 ||
+		    WT_CLOCKDIFF_SEC(now, fh->last_sync) > 0) {
+			/*
+			 * Skip over any tmp handles that are WiredTiger owned.
+			 * Adding a reference interferes with other internal
+			 * operations such as log archiving.
+			 */
+			while (fhtmp != NULL) {
+				ftp = strrchr(fhtmp->name, '/');
+				if (ftp == NULL)
+					ftp = fhtmp->name;
+				else
+					++ftp;
+				if (ftp[0] == 'W' &&
+				    WT_PREFIX_MATCH(ftp, "WiredTiger")) {
+					fhtmp = TAILQ_NEXT(fhtmp, q);
+					continue;
+				}
+				break;
+			}
+			/*
+			 * Increment our ref count on the current and next
+			 * handle. That way both are guaranteed valid when we
+			 * lock again after the fsync.
+			 */
+			++fh->ref;
+			if (fhtmp != NULL)
+				++fhtmp->ref;
+			__wt_spin_unlock(session, &conn->fh_lock);
+			ret = __wt_fsync(session, fh, false);
+			/*
+			 * Although we send in the false flag to indicate a
+			 * non-blocking background fsync, there is no guarantee
+			 * that it doesn't block. If we wanted to detect if it
+			 * is blocking, adding a clock call and checking the
+			 * time would be done here.
+			 */
+			/*
+			 * Even if we get an error, we still need to lock and
+			 * decrement our reference counts.
+			 */
+			if (ret == 0) {
+				F_CLR(fh, WT_FH_DIRTY);
+				WT_STAT_CONN_INCR(session, fsync_all_fh);
+				fh->last_sync = now;
+				WT_PUBLISH(fh->written, 0);
```

Comment:
Shouldn't we use `__wt_atomic_store64` instead of `WT_PUBLISH`?  Not sure about this, but `WT_PUBLISH` does a barrier operation before setting the value, so seems useful for forcing the order of operations, but does it guarantee the new value gets pushed out at this moment?

Reply:
I think the PUBLISH does imply pushing it out. And near this line of code but unrelated, I think we can get rid of the `WT_FH_DIRTY` flag. The `fh->written` value being non-zero implies the same thing.

Reply:
Having it `volatile` should make sure other threads see the value.